OpenRA.Mods.Common/Activities/DockActivity.cs

```diff
@@ -0,0 +1,161 @@
+				Cancel(self, true);
+
+			if (IsCanceling)
+				return true;
+
+			// Find the nearest best dock if not explicitly ordered to a specific dock
+			if (dock == null || dockManager.LinkedDock == null)
+				dock = dockManager.ChooseNewDock(null, desiredTypes);
+
+			// No docks exist; check again after delay defined in dockable.
+			if (dock == null || dockManager.LinkedDock == null)
+			{
+				var nearestDock = dockManager.ChooseNewDock(null, desiredTypes, true);
+				if (nearestDock != null)
+				{
+					var nearestDelta = (nearestDock.Position - self.CenterPosition).LengthSquared;
+					var dist = nearestDock.Info.WaitDistanceFromResupplyBase.Length;
+
+					// If no dock is available, move near one and wait
+					if (nearestDelta > dist)
+					{
+						var randomPosition = WVec.FromPDF(self.World.SharedRandom, 2) * dist / 1024;
+						var target = Target.FromPos(nearestDock.Position + randomPosition);
+
+						QueueChild(movement.MoveTo(nearestDock.Location, nearestDock.Info.CloseEnough < WDist.Zero ? 0 : nearestDock.Info.CloseEnough.Length, targetLineColor: dockManager.DockLineColor));
+					}
+					else
+						QueueChild(new Wait(dockManager.Info.SearchForDockDelay));
+				}
+				else
+					QueueChild(new Wait(dockManager.Info.SearchForDockDelay));
+
+				return false;
+			}
+
+			bool isCloseEnough;
+
+			// var delta = (dock.Position + (aircraft == null ? WVec.Zero : dock.Info.AircraftOffset) - self.CenterPosition).LengthSquared;
+			var delta = (dock.Position - self.CenterPosition).LengthSquared;
+
+			// Negative means there's no distance limit.
+			if (dock.Info.CloseEnough < WDist.Zero)
+				isCloseEnough = true;
+			else
+				isCloseEnough = self.Location == dock.Location;
+				// isCloseEnough = delta <= dock.Info.CloseEnough.LengthSquared;
+
+			if (!isCloseEnough)
+			{
+				foreach (var n in notifyDockActions)
+					n.MovingToDock(dockManager, dock);
+
+				// if (aircraft != null)
+				// {
+				// 	// QueueChild(movement.MoveWithinRange(Target.FromCell(self.World, dock.Location), WDist.Zero, targetLineColor: dockManager.DockLineColor));
+				// 	QueueChild(new Land(self, Target.FromCell(self.World, dock.Location), dock.Info.AircraftOffset, dock.Info.Facing, dockManager.DockLineColor));
+				// }
+				// else
+				QueueChild(movement.MoveTo(dock.Location, dock.Info.CloseEnough <= WDist.Zero ? 0 : dock.Info.CloseEnough.Length, targetLineColor: dockManager.DockLineColor));
+				transportCallers.FirstOrDefault(t => t.MinimumDistance.LengthSquared < delta)?.RequestTransport(self, dock.Location);
+
+				return false;
+			}
+
+			if (aircraft != null)
+				QueueChild(new Land(self, Target.FromActor(dock.Self), dock.Info.Facing, dockManager.DockLineColor));
+
+			QueueChild(new CallFunc(() => dock.DockedUnit = dockManager, false));
+			if (self.TraitOrDefault<WithVoxelUnloadBody>() == null)
+				QueueChild(new SpriteDockSequence(dockManager, dock));
+			else
+				QueueChild(new VoxelDockSequence(dockManager, dock));
```

Comment:
Sprite vs. Voxel distinction should be abstracted away from the docking code. Not sure how this should look precisely but calling a CreateDockSequence method from within a DockSequence class that handles this distinction makes sense to me.

Reply:
This PR is very out of date. I'll update it once some of the extracted PR's are merged

Reply:
I've already adressed this by removing TiberianSunRefinery